builtin/game/item_entity.lua

```diff
@@ -152,6 +178,111 @@ core.register_entity(":__builtin:item", {
 			return
 		end
 
+		local node_this = core.get_node_or_nil(pos)
+		local is_stuck = false
+		if node_this then
+			local sdef = minetest.registered_nodes[node_this.name]
+			is_stuck = (sdef.walkable == nil or sdef.walkable == true)
+				and (sdef.collision_box == nil or sdef.collision_box.type == "regular")
+				and (sdef.node_box == nil or sdef.node_box.type == "regular")
+		end
+
+		-- Push item out when stuck inside solid node
+		if is_stuck then
+			local shootdir
+			local cx = (pos.x % 1) - 0.5
+			local cz = (pos.z % 1) - 0.5
+			local order = {}
+
+			-- First prepare the order in which the 4 sides are to be checked.
+			-- 1st: closest
+			-- 2nd: other direction
+			-- 3rd and 4th: other axis
+			local cxcz = function(o, cw, one, zero)
+				if cw < 0 then
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+				else
+					table.insert(o, { [one]=-1, y=0, [zero]=0 })
+					table.insert(o, { [one]=1, y=0, [zero]=0 })
+				end
+				return o
+			end
+			if math.abs(cx) < math.abs(cz) then
+				order = cxcz(order, cx, "x", "z")
+				order = cxcz(order, cz, "z", "x")
+			else
+				order = cxcz(order, cz, "z", "x")
+				order = cxcz(order, cx, "x", "z")
+			end
+
+			-- Check which one of the 4 sides is free
+			for o=1, #order do
+				local nn = minetest.get_node(vector.add(pos, order[o])).name
+				local def_side = minetest.registered_nodes[nn]
+				if def_side and def_side.walkable == false and nn ~= "ignore" then
+					shootdir = order[o]
+					break
+				end
+			end
+			-- If none of the 4 sides is free, shoot upwards
+			if shootdir == nil then
+				shootdir = { x=0, y=1, z=0 }
+				local nn = minetest.get_node(vector.add(pos, shootdir)).name
+				if nn == "ignore" then
+					-- Do not push into ignore
+					return
+				end
+			end
+
+			-- Set new item moving speed accordingly
+			local newv = vector.multiply(shootdir, 3)
+			self.object:set_acceleration({x = 0, y = 0, z = 0})
+			self.object:set_velocity(newv)
+
+			disable_physics(self.object, self)
+
+			if shootdir.y == 0 then
+				self.force_out = newv
+				self.force_out_start = vector.floor(pos)
+				self.force_out_timer = 1
+			end
+			return
+		end
+
+		-- This code is run after the entity got a push from above “push away” code.
+		-- It is responsible for making sure the entity is entirely outside the solid node
+		-- (with its full collision box), not just its center.
+		if self.force_out_timer > 0 then
+			local c = self.object:get_properties().collisionbox
+			local s = self.force_out_start
+			local f = self.force_out
+			local ok = false
+			if f.x > 0 and (pos.x > (s.x + 0.5 + (c[4] - c[1])/2)) then
```

Comment:
This is overcomplicated and only works correctly for symmetric collisionboxes. Why do you do this?
```suggestion
			if f.x > 0 and pos.x > s.x + c[1] or
					f.x < 0 and pos.x < s.x + c[4] or
					... then
```